### DISCUSSION TITLE
Private npm registry for testing

### DISCUSSION THREAD

**Discussion by hisener** [renovatebot/renovate#8368] [Request Help] [answered]
Hey,

I am trying to reproduce a bug, but I cannot reproduce it against public registries. Is there a private registry that I can use for testing? So that I can create a reproducible example for the bug report.

Some background:

#### What Renovate type, platform and version are you using?

Self Hosted Renovate using 24.25.1-slim. 

`yarn.lock` updates fail consistently for some dependencies (e.g. typescript, react, etc.) but not for all. Locking for the same dependency works in npm's `package-lock.json`. I am suspicious about the npm cache but couldn't prove it yet. I am getting the following error:

```
File name: frontend/yarn.lock
error Couldn't find package "react" on the "npm" registry.
```

I have only `npmrc` in my config. Maybe that's the culprit. The relevant parts:

```
...
    npmrc: 'registry=https://...\n_auth=...',
    hostRules: [
        {
            hostName: 'registry.npmjs.org',
            enabled: false
        },
    ...
    ],
...
```

I also tried to set `dockerMapDotfiles: true`, but then I got 🤷‍♂️ :

```
File name: frontend/yarn.lock
error An unexpected error occurred: "EISDIR: illegal operation on a directory, read".
```

**Comment by rarkins** [comment, marked as answer]
I pay for npm and have a @rarkins/private-test package I use. I'm not sure if it's possible to grant another user access without them also needing to pay. 

I'd suggest Verdaccio as an option too, if you have a public server you can spin it up for.

**Reply by viceice** [reply]
Yeah, I'm using verdaccio too, very easy to install (docker) and simple configuration. 🙃

**Reply by juanpicado** [reply]
You can setup verdaccio in Digital Ocean in few minutes, expose the server and do your test, it will cost few cents for few hours, then shut it down. There is a how-to example at verdaccio blog using digital ocean.

**Reply by hisener** [reply]
Thanks, guys. I set up a Nexus instance using [docker image](https://hub.docker.com/r/sonatype/nexus3) and an npm proxy repo following [this guide](https://blog.sonatype.com/using-nexus-3-as-your-repository-part-2-npm-packages).

Turned out yarn is not sending the authentication header for all requests and this causes some 401s since the registry is private. Setting  `always-auth=true` in `npmrc` option resolved the issue. See https://github.com/yarnpkg/yarn/issues/3093#issuecomment-317671597 for more information.

**Comment by viceice** [comment]
Verdaccio:

* [docker](https://verdaccio.org/docs/en/docker)
* [kubernetes](https://verdaccio.org/docs/en/kubernetes) (with Helm)

🙃
